cdm/misc/src/test/java/ucar/nc2/geotiff/TestIFDEntry.java

```diff
@@ -0,0 +1,104 @@
+    result.add(new Object[] {new IFDEntry(null, FieldType.ASCII, 1), 0});
+    result.add(new Object[] {new IFDEntry(null, FieldType.ASCII, 1), Byte.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.ASCII, 1), 255});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SHORT, 1), 0});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SHORT, 1), Byte.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SHORT, 1), Short.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SHORT, 1), 65535});
+    result.add(new Object[] {new IFDEntry(null, FieldType.LONG, 1), 0});
+    result.add(new Object[] {new IFDEntry(null, FieldType.LONG, 1), Byte.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.LONG, 1), Short.MAX_VALUE});
+    // NOTE: because of the API design, unsigned longs can't be properly read or written
+    // for all possible values because Java's integer is signed.
+    result.add(new Object[] {new IFDEntry(null, FieldType.LONG, 1), Integer.MAX_VALUE});
+
+    // Signed
+    result.add(new Object[] {new IFDEntry(null, FieldType.SBYTE, 1), 0});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SBYTE, 1), Byte.MIN_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SBYTE, 1), Byte.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SSHORT, 1), 0});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SSHORT, 1), Byte.MIN_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SSHORT, 1), -Byte.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SSHORT, 1), Byte.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SSHORT, 1), Short.MIN_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SSHORT, 1), Short.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SLONG, 1), 0});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SLONG, 1), -Byte.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SLONG, 1), Byte.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SLONG, 1), -Short.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SLONG, 1), Short.MAX_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SLONG, 1), Integer.MIN_VALUE});
+    result.add(new Object[] {new IFDEntry(null, FieldType.SLONG, 1), Integer.MAX_VALUE});
+
+    return result;
+  }
+
+  public TestIFDEntry(IFDEntry ifd, int testValue) {
+    this.ifd = ifd;
+    this.testValue = testValue;
+  }
+
+  @Test
+  public void testRoundtrip() throws ReflectiveOperationException {
+    GeoTiff geotiff = new GeoTiff("foobar");
+    // 16 bytes should be more than enough
+    ByteBuffer buffer = ByteBuffer.allocate(16);
+    ByteOrder byteOrder = ByteOrder.BIG_ENDIAN;
+    buffer.order(byteOrder);
+
+    Method writeMethod =
```

Comment:
I personally don't like this reflection solution for testing private methods. First, in general, private methods should not be tested, though I understand that in practice some classes would need big refactors to make the public API unit testable. Second, I think it makes the test harder to read. And finally, if someone were to rename `writeIntValue`, the compiler would not catch an issue here, you would just get `NoSuchMethodException`.

Is there any reasonable way to test the public API of `GeoTiff` instead? I am not as familiar with the code, so I am not sure how feasible this is. If not, I propose to make the method `writeIntValue` package private so you can test it without reflection. Also not a perfect solution, but I think it is slightly nicer.

Reply:
Coming from python, the idea that one shouldn't test private methods is completely foreign to me. Unfortunately, with the way the GeoTiff class is designed, it would be difficult to not only construct the many geotiffs instances with the appropriate data types and ranges, but to then subsequently test that it was encoded correctly. The test would have a huge amount of unrelated setup that would make the test difficult to read and understand.

I am totally fine with changing it to be package private. We could even go a step further and make them static as they do not interact with any class members.